Userland/Libraries/LibWeb/DOM/Document.cpp

```diff
@@ -3767,4 +3767,46 @@ Element const* Document::element_from_point(double x, double y)
     return nullptr;
 }
 
+// https://drafts.csswg.org/cssom-view/#dom-document-elementsfrompoint
+Vector<JS::NonnullGCPtr<Element>> Document::elements_from_point(double x, double y)
+{
+    // 1. Let sequence be a new empty sequence.
+    Vector<JS::NonnullGCPtr<Element>> sequence;
+
+    // 2. If either argument is negative, x is greater than the viewport width excluding the size of a rendered scroll bar (if any),
+    //    or y is greater than the viewport height excluding the size of a rendered scroll bar (if any),
+    //    or there is no viewport associated with the document, return sequence and terminate these steps.
+    auto viewport_rect = this->viewport_rect();
+    CSSPixelPoint position { x, y };
+    // FIXME: This should account for the size of the scroll bar.
+    if (x < 0 || y < 0 || position.x() > viewport_rect.width() || position.y() > viewport_rect.height())
+        return sequence;
+
+    // Ensure the layout tree exists prior to hit testing.
+    update_layout();
+
+    // 3. For each box in the viewport, in paint order, starting with the topmost box, that would be a target for
+    //    hit testing at coordinates x,y even if nothing would be overlapping it, when applying the transforms that
+    //    apply to the descendants of the viewport, append the associated element to sequence.
+    // FIXME: Paintable box tree order is not the same as paint order. We need a helper to traverse the paint tree in
+    //        paint order with a custom callback.
+    if (auto const* paintable_box = this->paintable_box(); paintable_box) {
+        paintable_box->for_each_in_inclusive_subtree_of_type<Painting::PaintableBox>([&](auto& paintable_box) {
+            if (auto result = paintable_box.hit_test(position, Painting::HitTestType::Exact); result.has_value()) {
+                if (auto* dom_node = result->dom_node(); dom_node && dom_node->is_element())
+                    sequence.append(*static_cast<Element*>(dom_node));
+            }
+            return Painting::TraversalDecision::Continue;
+        });
+    }
+
+    // 4. If the document has a root element, and the last item in sequence is not the root element,
+    //    append the root element to sequence.
+    if (auto* document_root_element = first_child_of_type<Element>(); document_root_element)
```

Comment:
:thinking: can just use `document_element()`, right?

Reply:
it says in the CSS-DISPLAY-4 spec that the `root element` is

> The [element](https://drafts.csswg.org/css-display-4/#elements) at the root of the [document tree](https://dom.spec.whatwg.org/#concept-document-tree). In a document tree produced under the DOM, this is the [document element](https://dom.spec.whatwg.org/#document-element); in HTML it is the [html](https://html.spec.whatwg.org/multipage/semantics.html#the-html-element) element. [[DOM]](https://drafts.csswg.org/css-display-4/#biblio-dom) [[HTML]](https://drafts.csswg.org/css-display-4/#biblio-html)

 https://drafts.csswg.org/css-display-4/#root-element
 
 So it's... either the document element, or the html element?
 
 I honestly just copied this from the function above, so I'm not sure if it's correct or not either way...

Reply:
,,, oh that's literally what `document_element()` does. it calls `first_child_of_type<Element>()` :sweat_smile: